### PR TITLE
[feat/docs]: Documentation for range_queries/mo.cpp 

### PR DESCRIPTION
**Pull request by mishraabhinn**
#### Checklist

- [ ] Added description of change
- [ ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [ ] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] I acknowledge that all my contributions will be made under the project's license.
- [x] Relevant documentation/comments is changed or added
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/923"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/923"></a>

### CODE REVIEW

File: range_queries/mo.cpp [hunk 1]
```diff
@@ -1,50 +1,84 @@
...
+ * The preprocessing part takes O(m Log m) time.
+ * Processing all queries takes O(n * √n) + O(m * √n) = O((m+n) * √n) time.
+ */
+#include <algorithm>
 #include <iostream>
-using namespace std;
 const int N = 1e6 + 5;
 int a[N], bucket[N], cnt[N];
+// Variable to represent block size. This is made global so compare() of sort
+// can use it.
```
Comment (Panquesito7) [inline range_queries/mo.cpp:16]: Use a comment structure **except** inside functions.
Same in various other parts of the code.

Comment (mishraabhinn) [reply range_queries/mo.cpp:16]:
```
/*

*/
```

you mean to say this way.

Comment (Panquesito7) [reply range_queries/mo.cpp:16]: Yes, that would be correct. 👍

**Comment by kvedala** [comment]
good job on improving the code. Let us take this opportunity to fix as much as we can.
Reference: [orignial commit of the file](https://github.com/TheAlgorithms/C-Plus-Plus/commit/facfb2f35e2b3661f387afe5a0c805640aac49d9#diff-6af34138fbc1861feb654261aa0cdfa5) and I think it is supposed to be a solution for [this problem](https://www.spoj.com/problems/DQUERY/) and implements [Mo's algorithm](https://www.hackerearth.com/practice/notes/mos-algorithm/). [another reference](https://cp-algorithms.com/data_structures/sqrt_decomposition.html)
1. the filename `mo.cpp` is very ambiguous, can rename it to `mo_algorithm` by executing the following command from root folder:
```sh
git mv range_queries/mo.cpp range_queries/mo_algorithm.cpp
```
2. Please expand the description with above information
3. remove global variables

**Comment by mishraabhinn** [comment]
> good job on improving the code. Let us take this opportunity to fix as much as we can.
> Reference: [orignial commit of the file](https://github.com/TheAlgorithms/C-Plus-Plus/commit/facfb2f35e2b3661f387afe5a0c805640aac49d9#diff-6af34138fbc1861feb654261aa0cdfa5) and I think it is supposed to be a solution for [this problem](https://www.spoj.com/problems/DQUERY/) and implements [Mo's algorithm](https://www.hackerearth.com/practice/notes/mos-algorithm/). [another reference](https://cp-algorithms.com/data_structures/sqrt_decomposition.html)
> 
> 1. the filename `mo.cpp` is very ambiguous, can rename it to `mo_algorithm` by executing the following command from root folder:
> 
> ```shell
> $ git mv range_queries/mo.cpp range_queries/mo_algorithm.cpp
> $ git commit -m "file renamed"
> ```
> 
> 1. Please expand the description with above information
> 2. remove global variables

If global variable are removed.  We have to the supply them as argument in function definition,
Can we avoid this?
And by description do you mean to add in the more description in  the file.

**Comment by kvedala** [comment]
> > good job on improving the code. Let us take this opportunity to fix as much as we can.
> > Reference: [orignial commit of the file](https://github.com/TheAlgorithms/C-Plus-Plus/commit/facfb2f35e2b3661f387afe5a0c805640aac49d9#diff-6af34138fbc1861feb654261aa0cdfa5) and I think it is supposed to be a solution for [this problem](https://www.spoj.com/problems/DQUERY/) and implements [Mo's algorithm](https://www.hackerearth.com/practice/notes/mos-algorithm/). [another reference](https://cp-algorithms.com/data_structures/sqrt_decomposition.html)
> > 
> > 1. the filename `mo.cpp` is very ambiguous, can rename it to `mo_algorithm` by executing the following command from root folder:
> > 
> > ```shell
> > $ git mv range_queries/mo.cpp range_queries/mo_algorithm.cpp
> > $ git commit -m "file renamed"
> > ```
> > 
> > 
> > 
> > 1. Please expand the description with above information
> > 2. remove global variables
> 
> If global variable are removed. We have to the supply them as argument in function definition,
> Can we avoid this?
No. Using global variables are a seriously horrible coding practice (personal opinion) as it leads to so many bugs, and makes the code extremely vulnerable to attacks. You can try to run this program through `clang-tidy` to see what it reports back. 
> And by description do you mean to add in the more description in the file.
In my comment above, I provided with the background of the code, what it is supposed to achieve. If you can verify and incorporate the details into the file documentation, that would be great.

**Comment by Panquesito7** [comment]
@mishraabhinn any update here? Thanks.

**Comment by mishraabhinn** [comment]
Sorry @Panquesito7 and @kvedala  got busy in my things. I will try to do it once i get free.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/1796081447) warnings.

**Comment by Panquesito7** [comment]
> Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/1796081447) warnings.


Still failing.

File: range_queries/mo_algorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+#include <array>
+const int N = 1e6 + 5;
+std::array<int, N> a, bucket, cnt;
+// Variable to represent block size. This is made global so compare() of sort
+// can use it.
+int bucket_size;
+// Structure to represent a query range
+struct query {
+	int l, r, i;
+} q[N];
```
Comment (Panquesito7) [inline range_queries/mo_algorithm.cpp:23]:
```bash
/home/runner/work/C-Plus-Plus/C-Plus-Plus/range_queries/mo_algorithm.cpp:21:1: error: do not declare C-style arrays, use std::array<> instead [modernize-avoid-c-arrays,-warnings-as-errors].
```

**Review by Panquesito7 — COMMENTED** [review]
What is the `range_queries/mo_algorithm` file for?

File: range_queries/mo_algorithm.cpp (hunk 2 above)
Comment (mishraabhinn) [reply range_queries/mo_algorithm.cpp:23]: I used  std::array<int, N> q;
its throwing error.

Comment (Panquesito7) [reply range_queries/mo_algorithm.cpp:23]: Try using `std::vector` instead.

File: range_queries/mo_algorithm.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
+/**
+ *
```
Comment (Panquesito7) [inline range_queries/mo_algorithm.cpp:2]:
```suggestion
```

File: range_queries/mo_algorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+ * @file
+ * \brief A C++ Program to compute unique elements of ranges for different range
```
Comment (Panquesito7) [inline range_queries/mo_algorithm.cpp:4]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: range_queries/mo_algorithm.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+ * queries
+ *
+ * \details
+ * The preprocessing part takes O(m Log m) time.
+ * Processing all queries takes O(n * √n) + O(m * √n) = O((m+n) * √n) time.
```
Comment (Panquesito7) [inline range_queries/mo_algorithm.cpp:9]: Added "Unknown" author as I don't really know who created the original PR. You should add yourself, too, as you have improved the code a lot.
```suggestion
 * Processing all queries takes O(n * √n) + O(m * √n) = O((m+n) * √n) time.
 *
 * \author Unknown author
 * \author [Abhinn Mishra](https://github.com/mishraabhinn)
```

File: range_queries/mo_algorithm.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+ */
+#include <algorithm>
+#include <iostream>
+#include <cmath>
+#include <array>
+const int N = 1e6 + 5;
+std::array<int, N> a, bucket, cnt;
+// Variable to represent block size. This is made global so compare() of sort
+// can use it.
```
Comment (Panquesito7) [inline range_queries/mo_algorithm.cpp:18]: Remember, no *global* variables. Instead, use them in a class or as function parameters.

File: range_queries/mo_algorithm.cpp (hunk 2 above)
Comment (Panquesito7) [reply range_queries/mo_algorithm.cpp:23]: You're using `std::array` on line 16, but here you're using a C-style array.

File: range_queries/mo_algorithm.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	if (x.l / bucket_size != y.l / bucket_size) {
+		return x.l / bucket_size < y.l / bucket_size;
+	}
+	// Same block, sort by R value
+	return x.r < y.r;
+}
+
+/**
+ * Main Function
+ */
```
Comment (Panquesito7) [inline range_queries/mo_algorithm.cpp:56]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
All functions are missing proper documentation (see the example below).

```c++
    /**
    * @brief Utility function to print matrix
    * @tparam n number of matrix size
    * @param board matrix where numbers are saved
    * @returns void
    */
    template <size_t n>
    void printSolution(const std::array<std::array<int, n>, n> &board) {
      std::cout << "\n";
      for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
          std::cout << "" << board[i][j] << " ";
        }
        std::cout << "\n";
      }
    }
```

Please remove the `mo_algorithm` file, too (there's an extra file without the `.cpp` file format, and I don't think it's needed).

**Comment by Panquesito7** [comment]
I'm reworking this PR. Thanks. 🙂

**Pull request closed without merging**